smallrye-reactive-messaging-kafka/src/main/java/io/smallrye/reactive/messaging/kafka/fault/KafkaDeadLetterQueue.java

```diff
@@ -96,8 +98,36 @@ private String getThrowableMessage(Throwable throwable) {
 
     @Override
     public <K, V> CompletionStage<Void> handle(
-            IncomingKafkaRecord<K, V> record, Throwable reason) {
-        KafkaProducerRecord<K, V> dead = KafkaProducerRecord.create(topic, record.getKey(), record.getPayload());
+            IncomingKafkaRecord<K, V> record, Throwable reason, Metadata metadata) {
+
+        OutgoingKafkaRecordMetadata<K> outgoing = null;
+        if (metadata != null) {
+            for (Object item : metadata) {
+                if (item instanceof OutgoingKafkaRecordMetadata) {
+                    outgoing = (OutgoingKafkaRecordMetadata<K>) item;
+                    // more than 1 instance => undefined behavior
```

Comment:
actually, not supported by the metadata model.

Reply:
We can add an accessor by type in `Metadata`:
```
Optional<OutgoingKafkaRecordMetadata> maybe = metadata.get(OutgoingKafkaRecordMetadata.class)
```

Reply:
Yea I noticed that `Message` has an accessor like this. Frankly, I don't see why `Metadata` are not backed by a `Map`, because what we're currently doing (see e.g. `replaceOrAdd`) is weird and inefficient (and, in case of `replaceOrAdd`, also wrong :-) ).

Let me take a crack at this.

Reply:
Added a commit with refactored `Metadata` implementation (backing store is now a `Map<Class<?>, Object>`), and a `get(Class)` method.

Note that the signature of the `get` method I added is `<T> Optional<T> get(Class<T> clazz)`, so it looks up an object that is precisely of given class. The `Message.getMetadata(Class)` method is `<M> Optional<M> getMetadata(Class<? extends M> clazz)`, so it looks up an object that is of a subtype of given class. It looks strange to me, was that even intentional?